crates/wast/src/lexer.rs

```diff
@@ -506,236 +612,180 @@ impl<'a> Lexer<'a> {
         }
     }
 
-    /// Attempts to consume whitespace from the input stream, returning `None`
-    /// if there's no whitespace to consume
-    fn ws(&mut self) -> Option<&'a str> {
-        let start = self.cur();
-        loop {
-            match self.it.peek() {
-                Some((_, ' ')) | Some((_, '\n')) | Some((_, '\r')) | Some((_, '\t')) => {
-                    drop(self.it.next())
-                }
-                _ => break,
-            }
-        }
-        let end = self.cur();
-        if start != end {
-            Some(&self.input[start..end])
-        } else {
-            None
+    /// Verifies that `comment`, which is about to be returned, has a "confusing
+    /// unicode character" in it and should instead be transformed into an
+    /// error.
+    fn check_confusing_comment(&self, comment: &str) -> Result<(), Error> {
+        if self.allow_confusing_unicode {
+            return Ok(());
         }
-    }
 
-    /// Attempts to read a comment from the input stream
-    fn comment(&mut self) -> Result<Option<Token<'a>>, Error> {
-        if let Some(start) = self.eat_str(";;") {
-            let mut it = self.it.clone();
-            loop {
-                match it.next() {
-                    None | Some((_, '\n')) => break,
-                    Some((i, ch)) => {
-                        self.check_if_confusing(i, ch)?;
-                        self.it = it.clone();
+        // In an effort to avoid utf-8 decoding the entire `comment` the search
+        // here is a bit more optimized. This checks for the `0xe2` byte because
+        // in the utf-8 encoding that's the leading encoding byte for all
+        // "confusing characters". Each instance of 0xe2 is checked to see if it
+        // starts a confusing character, and if so that's returned.
```

Comment:
Similar to above, it might be helpful to note here that `0xe2` can only appear as the first byte of a character (all continuing bytes in a utf-8 sequence are `0x80..0xbf`) so there is no need to consider the case that we catch an `0xe2` in the middle of a codepoint.